temporal-sdk/src/main/java/io/temporal/internal/nexus/PayloadSerializer.java

```diff
@@ -52,7 +52,10 @@ public Content serialize(@Nullable Object o) {
   public @Nullable Object deserialize(Content content, Type type) {
     try {
       Payload payload = Payload.parseFrom(content.getData());
-      return dataConverter.fromPayload(payload, type.getClass(), type);
+      if (!(type instanceof Class)) {
```

Comment:
No, this expectation wouldn't extend to activities. Activities and workflows implementations can support some level of generic types

Reply:
by generics do you mean the activity or workflow interface is generic, or the input/out is a generic type?

Reply:
When the SDK has to decode a payload it takes the `Type` and `Class<R>` as separate parameters https://github.com/temporalio/sdk-java/blob/ac3526b711bdabdfce8d3a8aa35929360bf5e1fc/temporal-sdk/src/main/java/io/temporal/client/WorkflowStub.java#L206

Reply:
> by generics do you mean the activity or workflow interface is generic, or the input/out is a generic type?

In/out. Just curious why input/output for workflows/activities is more flexible than here and/or why we can't make them consistent. No strong opinion either way though.

Reply:
>Or maybe because we don't offer the same type of class + type overload for getting a result?

This, I am limited by the java nexus SDK interface. I think `Serializer` would need to be a generic interface to make it consistent

Reply:
Hrmm, I suppose I figured if it just accepted `Type` that'd be enough for generic types. Seeing https://github.com/uber/cadence-java-client/pull/179, I wonder if there wasn't a `Class` version before if just one accepting `Type` would be good enough.

Reply:
No `Type ` has no information about the generic type used

Reply:
Hrmm, I would have expected if it wasn't a `Class` and it was something like `ParameterizedType`, you could `getRawType`. I think we need to support things like `List<String>` and such with Nexus Java, but I figured APIs (Java reflection, Jackson, and our own SDK) only offer `Class` forms for legacy reasons. For instance, can Jackson deserialize if I just called `mapper.getTypeFactory().constructType(new TypeToken<List<String>>() {}.getType())` instead of `mapper.getTypeFactory().constructType(mapper.getTypeFactory().constructType(new TypeToken<List<String>>() {}.getType()), List.class)`? The latter is deprecated in Jackson presumably because the former is enough.

I think Nexus needs to support those generics, but I am hesitant to add a two-arg deserialize overload if the only reason is because the Temporal data converter needs it.

Reply:
Since the well I think you need to use `getGenericParameterTypes` instead of `getParameterTypes` no?

Reply:
Yeah, looking at it, it's probably a bug. Not sure why I did `getGenericReturnType()` but not `getGenericParameterTypes()`. Probably just lack of tests for generics (I regrettably didn't test very much at all). I think end-to-end we should support `List<Foo>` as a deserialized parameter for the imlpementer or as a deserialized return type for the caller.

Reply:
OK I updated this PR based off what I did here https://github.com/nexus-rpc/sdk-java/pull/17

Reply:
Yes, the data converter interface requires it `fromPayload(Payload payload, Class<T> valueClass, Type valueType)`

Reply:
Yeah I added a test for `HashMap` which I think is similar to testing `List`

Reply:
> Yeah I added a test for HashMap which I think is similar to testing List

Can you have one specifically for a `HashMap` that doesn't use primitives? Basically can we change:

```java
    Map<String, Integer> map = Collections.singletonMap("key", 0);
    PayloadSerializer.Content content = payloadSerializer.serialize(map);
    payloadSerializer.deserialize(content, map.getClass());
    Assert.assertEquals(map, payloadSerializer.deserialize(content, map.getClass()));
```

to something like:

```java
    Map<String, Foo> map = Collections.singletonMap("key", new Foo());
    PayloadSerializer.Content content = payloadSerializer.serialize(map);
    payloadSerializer.deserialize(content, (new TypeToken<Map<String, Foo>>() {}).getType());
    Assert.assertTrue(map.get("key") instanceof Foo);
```

I see a similar test at `EncodedValuesTest::testGenericParameter` and some other places using `TypeToken`. Can also have a test that does more end-to-end with some generic where the generic type is a specific type.

Reply:
Thanks for this test, it found a bug in my implementation. I thought testing `String` covered this case but it was not